lib/elsa/producer.ex

```diff
@@ -27,63 +29,90 @@ defmodule Elsa.Producer do
+  defp transform_message(message), do: {"", message}
+
+  defp do_produce_sync(topic, messages, opts) do
+    do_with_valid_client(opts, fn client ->
+      partitioner = get_partitioner(client, topic, opts)
+      message_chunks = create_message_chunks(partitioner, messages)
 
-    do_produce_sync(name, topic, partition, [{"", message}])
+      case produce_sync_while_successful(client, topic, message_chunks) do
+        {:ok, _} -> :ok
+        {:error, reason, chunks_sent} -> failure_message(message_chunks, reason, chunks_sent)
+      end
+    end)
   end
 
-  defp transform_message({key, value}, name, topic, opts) do
-    {get_partition(name, topic, key, opts), {key, value}}
+  defp do_with_valid_client(opts, function) when is_function(function, 1) do
+    case get_valid_client(opts) do
+      {:ok, client} -> function.(client)
+      error -> error
+    end
   end
 
-  defp transform_message(message, name, topic, opts) do
-    {get_partition(name, topic, "", opts), {"", message}}
+  defp produce_sync_while_successful(client, topic, message_chunks) do
+    Enum.reduce_while(message_chunks, {:ok, 0}, fn {partition, chunk}, {:ok, chunks_sent} ->
+      total_size = Enum.reduce(chunk, 0, fn {key, value}, acc -> acc + byte_size(key) + byte_size(value) end)
+
+      Logger.debug(fn ->
+        "#{__MODULE__} Sending #{length(chunk)} messages to #{topic}:#{partition} - Size : #{total_size}"
+      end)
+
+      case :brod.produce_sync(client, topic, partition, "", chunk) do
+        :ok -> {:cont, {:ok, chunks_sent + 1}}
+        {:error, reason} -> {:halt, {:error, reason, chunks_sent}}
+      end
+    end)
   end
 
-  defp do_produce_sync(name, topic, partition, messages) do
+  defp create_message_chunks(partitioner, messages) do
     messages
-    |> Enum.map(&wrap_with_key/1)
-    |> Util.chunk_by_byte_size()
-    |> Enum.each(fn chunk -> :brod.produce_sync(name, topic, partition, "", chunk) end)
+    |> Enum.group_by(partitioner)
+    |> Enum.map(fn {partition, messages} -> {partition, Util.chunk_by_byte_size(messages)} end)
+    |> Enum.flat_map(fn {partition, chunks} -> Enum.map(chunks, fn chunk -> {partition, chunk} end) end)
   end
 
-  defp wrap_with_key({_key, _value} = message), do: message
-  defp wrap_with_key(message), do: {"", message}
+  defp failure_message(message_chunks, reason, chunks_sent) do
+    messages_sent = Enum.take(message_chunks, chunks_sent) |> Enum.flat_map(fn {_partition, chunk} -> chunk end)
+
+    reason_string =
+      "#{length(messages_sent)} messages succeeded before elsa producer failed midway through due to #{inspect(reason)}"
+
+    failed_messages = Enum.drop(message_chunks, chunks_sent) |> Enum.flat_map(fn {_partition, chunk} -> chunk end)
+    {:error, reason_string, failed_messages}
+  end
 
   defp get_client(opts) do
-    Keyword.get_lazy(opts, :name, &Elsa.default_client/0)
+    Keyword.get_lazy(opts, :client, fn -> Keyword.get_lazy(opts, :name, &Elsa.default_client/0) end)
```

Comment:
It feels weird to have two nested lazy gets, but I guess it makes sense if you're trying to create an override order/ compatibility for client or name....